test/c/test_c.py

```diff
@@ -0,0 +1,271 @@
+import os
+import re
+import subprocess
+import pytest
+import shutil
+from pathlib import Path
+from typing import List, Optional
+
+
+def _get_sail_dir() -> Path:
+    from_env = os.environ.get("SAIL_DIR")
+    if from_env is not None:
+        return Path(from_env)
+    return Path(__file__).absolute().parent.parent
+
+
+def _get_sail() -> Path:
+    sail_bin = shutil.which(os.environ.get("SAIL", "sail"))
+    return sail_bin if sail_bin is not None else Path("sail")
+
+
+def _have_valgrind() -> bool:
+    try:
+        subprocess.check_call(["valgrind", "--version"])
+        return True
+    except subprocess.CalledProcessError:
+        return False
+
+
+SAIL_DIR = _get_sail_dir()
+SAIL = _get_sail()
+HAVE_VALGRIND = _have_valgrind()
+
+
+def find_sail_files() -> List[pytest.param]:
+    return [
+        pytest.param(test, id=test.name)
+        for test in Path(__file__).parent.glob("*.sail")
+    ]
+
+
+def run_command(
+    command: List[str],
+    expected_status: int = 0,
+    stderr_file: Optional[Path] = None,
+    cwd: Optional[Path] = None,
+) -> str:
+    process = subprocess.run(
+        command, stdout=subprocess.PIPE, stderr=subprocess.PIPE, cwd=cwd
+    )
+    if process.returncode != expected_status:
+        pytest.fail(
+            f"Command '{' '.join(command)}' failed with exit code {process.returncode}\nstdout:\n{process.stdout.decode()}\nstderr:\n{process.stderr.decode()}"
+        )
+    if stderr_file:
+        with open(stderr_file, "wb") as f:
+            f.write(process.stderr)
+    return process.stdout.decode()
+
+
+def _run_sail_c_test(
+    sail_file: Path,
+    tmp_path: Path,
+    sail_opts: List[str],
+    c_opts=None,
+    valgrind: bool = False,
+    compiler: str = "cc",
+    actually_cpp: bool = False,
+    expected_failures=None,
+) -> None:
+    if expected_failures is None:
+        expected_failures = {}
+    if c_opts is None:
+        c_opts = []
+    if sail_file.name in expected_failures:
+        pytest.xfail(expected_failures[sail_file.name])
+    sail_file_abs = sail_file.absolute()
+    basename = sail_file.stem
+
+    if actually_cpp:
+        extension = "cpp"
+    else:
+        extension = "c"
+
+    if not os.path.isfile(SAIL) or not os.access(SAIL, os.X_OK):
+        pytest.fail(f"Sail binary not found or not executable at {SAIL}")
+
+    # Run from the tests directory to avoid embedding an absolute path to the sail
+    # file in the error output.
+    run_command(
+        [SAIL] + sail_opts + [sail_file.name, "-o", tmp_path / basename],
+        cwd=sail_file_abs.parent,
+    )
+
+    lib_files = [str(f) for f in Path(SAIL_DIR, "lib").glob("*.c")]
+
+    run_command(
+        [compiler]
+        + c_opts
+        + [f"{basename}.{extension}"]
+        + lib_files
+        + ["-lgmp", "-I", f"{SAIL_DIR}/lib", "-o", f"{basename}.bin"],
+        cwd=tmp_path,
+    )
+
+    # For files using a config file, copy the config file to the temporary directory.
```

Comment:
This seems a bit hacky and it might be possible to just run the test with the CWD set to the tests directory so it can find the config json file. I did notice that test/c/config_abstract_type2.sail loads the "wrong" config file, not sure if that is intentional?